src/Core/Algorithm/Subdivision/CatmullClarkSubdivider.cpp

```diff
@@ -0,0 +1,393 @@
+#include <Core/Algorithm/Subdivision/CatmullClarkSubdivider.hpp>
+#include <Core/Algorithm/Subdivision/SubdividerUtils.hpp>
+
+namespace Ra {
+namespace Core {
+
+bool CatmullClarkSubdivider::prepare( TopologicalMesh& mesh ) {
+    mesh.add_property( m_vpPos );
+    mesh.add_property( m_epPos );
+    mesh.add_property( m_fpPos );
+    mesh.add_property( m_creaseWeights );
+    addPropsCopy( mesh.m_floatPph, mesh, m_floatProps );
+    addPropsCopy( mesh.m_vec2Pph, mesh, m_vec2Props );
+    addPropsCopy( mesh.m_vec3Pph, mesh, m_vec3Props );
+    addPropsCopy( mesh.m_vec4Pph, mesh, m_vec4Props );
+    addPropsCopyF( mesh.m_floatPph, mesh, m_floatPropsF );
+    addPropsCopyF( mesh.m_vec2Pph, mesh, m_vec2PropsF );
+    addPropsCopyF( mesh.m_vec3Pph, mesh, m_vec3PropsF );
+    addPropsCopyF( mesh.m_vec4Pph, mesh, m_vec4PropsF );
+
+    // initialize all weights to 0 (= smooth edge)
+    for ( auto e_it = mesh.edges_begin(); e_it != mesh.edges_end(); ++e_it )
+        mesh.property( m_creaseWeights, *e_it ) = 0.0;
+
+    return true;
+}
+
+bool CatmullClarkSubdivider::cleanup( TopologicalMesh& mesh ) {
+    mesh.remove_property( m_vpPos );
+    mesh.remove_property( m_epPos );
+    mesh.remove_property( m_fpPos );
+    mesh.remove_property( m_creaseWeights );
+    clearProps( m_floatProps, mesh );
+    clearProps( m_vec2Props, mesh );
+    clearProps( m_vec3Props, mesh );
+    clearProps( m_vec4Props, mesh );
+    clearProps( m_floatPropsF, mesh );
+    clearProps( m_vec2PropsF, mesh );
+    clearProps( m_vec3PropsF, mesh );
+    clearProps( m_vec4PropsF, mesh );
+    return true;
+}
+
+bool CatmullClarkSubdivider::subdivide( TopologicalMesh& mesh, size_t n,
+                                        const bool update_points ) {
+    // Do n subdivisions
+    for ( size_t i = 0; i < n; ++i )
+    {
+        // Compute face centroid
+#pragma omp parallel for
```

Comment:
We need to protect this with a parameter set when calling the method or constructing the `CatmullClarkSubdivider`. Indeed, in some context we may not want to have this part of the code parallelized.

Reply:
Why should we? Could you explain such a case?

Reply:
If I recall correctly, your point here was that at one time, OpenMP didn't handle nested parallelized loops, which he now does.